### PR TITLE
Implementation for dateStyle&timeStyle to Intl::DateTimeFormat

### PR DESCRIPTION
**Pull request by mganandraj**
This PR is adding implementation for dateStyle and timeStyle options to Intl:DateTimeFormat service for Android API 24+.

Issue No : https://github.com/facebook/hermes/issues/926

## Summary

This PR is adding implementation for dateStyle and timeStyle options to Intl:DateTimeFormat service.

## Test Plan
test262 based test suite is passing.

### CODE REVIEW

File: lib/Platform/Intl/java/com/facebook/hermes/intl/DateTimeFormat.java [hunk 1]
```diff
@@ -124,6 +127,17 @@ private Object ToDateTimeOptions(Object options, String required, String default
       }
     }
 
+    if (!JSObjects.isUndefined(JSObjects.Get(options, "dateStyle")) || !JSObjects.isUndefined(JSObjects.Get(options, "timeStyle")) )
+      needDefaults = false;
+
+    if (required.equals("date") && !JSObjects.isUndefined(JSObjects.Get(options, "timeStyle"))) {
+      // throw // TODO throw TypeError
```
Comment (jpporto) [inline lib/Platform/Intl/java/com/facebook/hermes/intl/DateTimeFormat.java:134]: can you resolve this todo?

File: lib/Platform/Intl/java/com/facebook/hermes/intl/DateTimeFormat.java [hunk 2]
```diff
@@ -124,6 +127,17 @@ private Object ToDateTimeOptions(Object options, String required, String default
...
+    }
+
+    if (required.equals("time") && !JSObjects.isUndefined(JSObjects.Get(options, "dateStyle"))) {
+      // throw // TODO throw TypeError
```
Comment (jpporto) [inline lib/Platform/Intl/java/com/facebook/hermes/intl/DateTimeFormat.java:138]: can you resolve this todo?

File: lib/VM/JSLib/Intl.cpp [hunk 3]
```diff
@@ -386,6 +387,34 @@ ExecutionStatus checkOptions(Runtime &runtime, const platform_intl::Options &) {
   return ExecutionStatus::RETURNED;
 }
 
+template <>
+ExecutionStatus checkOptions<platform_intl::DateTimeFormat>(
+    Runtime &runtime,
+    const platform_intl::Options &options) {
+  // https://www.ecma-international.org/wp-content/uploads/ECMA-402_9th_edition_june_2022.pdf
+  //  InitializeDateTimeFormat # 42
+  const bool hasExplicitFormatComponents {
```
Comment (jpporto) [inline lib/VM/JSLib/Intl.cpp:396]: please use = instead of {}

File: lib/VM/JSLib/Intl.cpp [hunk 4]
```diff
@@ -434,7 +463,6 @@ CallResult<HermesValue> intlServiceConstructor(
     return ExecutionStatus::EXCEPTION;
   }
 
-  // NumberFormat has a couple extra checks to make.
```
Comment (jpporto) [inline lib/VM/JSLib/Intl.cpp:437]: why is the comment gone?

File: lib/VM/JSLib/Intl.cpp [hunk 5]
```diff
@@ -1424,6 +1452,14 @@ CallResult<HermesValue> intlDatePrototypeToSomeLocaleString(
     }
     toDateTimeOptions(*optionsRes, dtoFlags);
 
+    if (!(dtoFlags & kDTOTime) && (*optionsRes).find(u"timeStyle") != (*optionsRes).end())
```
Comment (jpporto) [inline lib/VM/JSLib/Intl.cpp:1455]: use optionsRes-> instead of (*optionsRes).

File: lib/Platform/Intl/java/com/facebook/hermes/intl/DateTimeFormat.java (hunk 1 above)
Comment (mganandraj) [reply lib/Platform/Intl/java/com/facebook/hermes/intl/DateTimeFormat.java:134]: This check is added in the native code .. I initially added this TODO in the java code to not forget adding checks to native code ..
Currently we don't have the mechanism to throw the validation errors from Java code .. all such validations must be done in C++ code today..

File: lib/Platform/Intl/java/com/facebook/hermes/intl/DateTimeFormat.java (hunk 2 above)
Comment (mganandraj) [reply lib/Platform/Intl/java/com/facebook/hermes/intl/DateTimeFormat.java:138]: Same as above ..

File: lib/VM/JSLib/Intl.cpp (hunk 4 above)
Comment (mganandraj) [reply lib/VM/JSLib/Intl.cpp:437]: This function is a generic one for all the services .. not just NumberFormat .. But only NumberFormat had a service specific checks made earlier.. but this PR introduces DataFormat specific checks .. 
May be this comment can be made generic ..

File: lib/VM/JSLib/Intl.cpp [hunk 6]
```diff
@@ -1424,6 +1452,14 @@ CallResult<HermesValue> intlDatePrototypeToSomeLocaleString(
     }
     toDateTimeOptions(*optionsRes, dtoFlags);
```
Comment (jpporto) [inline lib/VM/JSLib/Intl.cpp:1453]: It took me a while to understand what's going on here. These added checks are defined as part of [toDateTimeOptions](https://tc39.es/ecma402/#sec-todatetimeoptions), so please make them on that function. It also seems that steps 6., 7., and 8. are missing from the implementation (i.e., if either ```timeStyle``` or ```dateStyle``` are present, then ```needsDefault = false```).

File: lib/VM/JSLib/Intl.cpp [hunk 7]
```diff
@@ -386,6 +387,33 @@ ExecutionStatus checkOptions(Runtime &runtime, const platform_intl::Options &) {
   return ExecutionStatus::RETURNED;
 }
 
+template <>
+ExecutionStatus checkOptions<platform_intl::DateTimeFormat>(
+    Runtime &runtime,
+    const platform_intl::Options &options) {
+  // https://www.ecma-international.org/wp-content/uploads/ECMA-402_9th_edition_june_2022.pdf
```
Comment (jpporto) [inline lib/VM/JSLib/Intl.cpp:394]: The style in this file for quoting the spec is

```
// ECMA 402 11.1.2
35. Let hasExplicitFormatComponents be false.
36. For each row of table https://tc39.es/ecma402/#table-datetimeformat-components, except the header row, in table order, do
    e. If value is not undefined, then
        i. Set hasExplicitFormatComponents to true.
```

Perhaps add a 2023 to the first line, i.e.,

```
// ECMA 402 2023 11.1.2
```

File: lib/VM/JSLib/Intl.cpp [hunk 8]
```diff
@@ -386,6 +387,33 @@ ExecutionStatus checkOptions(Runtime &runtime, const platform_intl::Options &) {
...
+      options.find(u"day") != options.end() ||
+      options.find(u"day") != options.end() ||
+      options.find(u"dayPeriod") != options.end() ||
+      options.find(u"hour") != options.end() ||
+      options.find(u"minute") != options.end() ||
+      options.find(u"second") != options.end() ||
+      options.find(u"fractionalSecondDigits") != options.end() ||
+      options.find(u"timeZoneName") != options.end();
+  const bool hasStyle = options.find(u"dateStyle") != options.end() || options.find(u"timeStyle") != options.end();
+  if ( hasStyle && hasExplicitFormatComponents) {
```
Comment (jpporto) [inline lib/VM/JSLib/Intl.cpp:409]: Add a comment

```
// 42. If dateStyle is not undefined or timeStyle is not undefined, then
//    a. If hasExplicitFormatComponents is true, then
//        i. Throw a TypeError exception.
```

before the if to link the behavior with the Spec section defining it.

File: lib/VM/JSLib/Intl.cpp [hunk 9]
```diff
@@ -386,6 +387,33 @@ ExecutionStatus checkOptions(Runtime &runtime, const platform_intl::Options &) {
...
+    Runtime &runtime,
+    const platform_intl::Options &options) {
+  // https://www.ecma-international.org/wp-content/uploads/ECMA-402_9th_edition_june_2022.pdf
+  //  InitializeDateTimeFormat # 42
+  const bool hasExplicitFormatComponents = options.find(u"weekday") != options.end() ||
+      options.find(u"era") != options.end() ||
+      options.find(u"year") != options.end() ||
+      options.find(u"month") != options.end() ||
+      options.find(u"day") != options.end() ||
+      options.find(u"day") != options.end() ||
```
Comment (jpporto) [inline lib/VM/JSLib/Intl.cpp:401]: duplicate check for day

File: lib/VM/JSLib/Intl.cpp [hunk 10]
```diff
@@ -386,6 +387,33 @@ ExecutionStatus checkOptions(Runtime &runtime, const platform_intl::Options &) {
   return ExecutionStatus::RETURNED;
 }
 
+template <>
+ExecutionStatus checkOptions<platform_intl::DateTimeFormat>(
+    Runtime &runtime,
+    const platform_intl::Options &options) {
+  // https://www.ecma-international.org/wp-content/uploads/ECMA-402_9th_edition_june_2022.pdf
+  //  InitializeDateTimeFormat # 42
+  const bool hasExplicitFormatComponents = options.find(u"weekday") != options.end() ||
```
Comment (jpporto) [inline lib/VM/JSLib/Intl.cpp:396]: prefer using ```.count(<option>);``` so there's no need to call ```.end()``` several times.

File: lib/VM/JSLib/Intl.cpp [hunk 11]
```diff
@@ -287,10 +288,10 @@ constexpr OptionData kDTFOptions[] = {
     {u"second",
      platform_intl::Option::Kind::String,
      kTimeRequired | kTimeDefault},
+    {u"fractionalSecondDigits", platform_intl::Option::Kind::Number, 0},
```
Comment (jpporto) [inline lib/VM/JSLib/Intl.cpp:291]: just curious: is this change needed?

File: lib/VM/JSLib/Intl.cpp [hunk 12]
```diff
@@ -434,7 +464,7 @@ CallResult<HermesValue> intlServiceConstructor(
     return ExecutionStatus::EXCEPTION;
   }
 
-  // NumberFormat has a couple extra checks to make.
+  // Service specific checks ..
```
Comment (jpporto) [inline lib/VM/JSLib/Intl.cpp:467]: two ..?

File: lib/VM/JSLib/Intl.cpp [hunk 13]
```diff
@@ -1395,15 +1425,27 @@ void toDateTimeOptions(platform_intl::Options &options, int dtoFlags) {
...
-    if ((dtoFlags & kDTODate && pod.flags & kDateDefault) ||
-        (dtoFlags & kDTOTime && pod.flags & kTimeDefault)) {
-      options.emplace(pod.name, std::u16string(u"numeric"));
+
+  const bool hasStyle = options.count(u"dateStyle") > 0 || options.count(u"timeStyle") > 0;
+  if (hasStyle)
+    needDefaults = false;
+
+  if (!(dtoFlags & kDTOTime) && options.count(u"timeStyle") > 0)
+    return ExecutionStatus::EXCEPTION;
```
Comment (jpporto) [inline lib/VM/JSLib/Intl.cpp:1434]: this is not how we use ```ExecutionStatus```. You should modify this function to take a Runtime, and then use ```Runtime::raiseTypeError``` here.

**Review by jpporto — COMMENTED** [review]
This should be good to land once you change the ```ExecutionStatus``` usage in toDateTimeOptions

File: lib/VM/JSLib/Intl.cpp (hunk 11 above)
Comment (mganandraj) [reply lib/VM/JSLib/Intl.cpp:291]: It is not needed .. but thought it would be nice to keep it alongside rest of the time fields.

File: lib/VM/JSLib/Intl.cpp (hunk 13 above)
Comment (mganandraj) [reply lib/VM/JSLib/Intl.cpp]: Absolutely ..

**Comment by nandorojo** [comment]
Hey, why is this closed? Did it get merged elsewhere? Is there `dateStyle`/`timeStyle` support now?

**Comment by tmikov** [comment]
@nandorojo if you click on the commit where it was merged https://github.com/facebook/hermes/commit/a72115fb24414fde85e5fe5e29085431205454d0 it shows you which branches contain that commit.
